webscrapping101/scrape.markdown

```diff
@@ -0,0 +1,313 @@
+
+```python
+import httpx
+
+url = "https://bulletin.engin.umich.edu/courses/eecs/"
+headers = {
+    'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8',
+    'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.97 Safari/537.36',
+}
+response = httpx.get(url,headers=headers)
+print(response.text)
+```
+
+Running the code above, we get the following output. (Note that all the code are in scrape.ipynb for you to try out on your own)
+
+![out1](./image/out1.png)
+
+The output is the html content of the webpage
+
+We can analyze this returned HTML directly, or even better, use developer tools provided by browser to analyze the website's components, layout, and structures.
+
+Open up "Inspect Element" in Chrome. (This can be done by clicking on "View" -> "Developer" -> "Inspect Element")
+
+![out2](./image/umich2.png)
+
+Through some examination, we can see that each of the course information are enclosed in a "\<p>" tag under "\<div class=entry-content>". The course code and name are in the "\<strong>" tag. The course credit and prerequisite are in "\<em>", and finally, the atlas link is embedded in a "\<a href>" hyperlink tag. 
+
+Here is a more zoomed in view. 
+![](./image/zoom_in_course.png)
+
+We want to parse all of the informations enclosed in tags listed above for every course under the p tag. We will do this using Beautifulsoup. 
+
+and here is a zoomed out view of all the course blocks. 
+![](./image/umich3.png)
+
+
+First take a look of selecting all the p tags under "\<div class="entry-content">" and printing their cotent out in full. 
+
+```python
+from bs4 import BeautifulSoup
+
+parser = BeautifulSoup(response, 'html.parser')
+
+div = parser.find('div', class_= "entry-content")
+
+p_tags = div.find_all('p')[1:] # through inspection the first tag is irrelevant.
+
+for p in p_tags:
+    print(p.get_text())
+```
+
+![](./image/p_clean.png)
+
+A raw look
+
+
+```python
+for p in p_tags:
+    print(p)
+```
+
+![](./image/p_raw.png)
+
+
+As described in our guide, we only need the informations under the "strong", "em", and "a" tag. For "a" tag, we need its "href" attribute for a web url (instead of the literal words "CourseProfile (ATLAS)" )
+
+```python 
+course = []
+prereq = []
+link = []
+
+for p in p_tags:
+    strong, em, a = p.find('strong'), p.find('em'),  p.find('a')  
+    if strong is not None: course.append(strong.getText())
+    if em is not None: prereq.append(em.getText())
+    if a is not None: link.append(a['href']) # Get the href attribute, not literal text
+
+print(course)
+print(prereq)
+print(link)
+
+```
+![](./image/p_2.png)
+
+Some of the courses have the character \xa0, a variant of a standard space chraracter that prevents an automatic link break at its position. We need to strip it. 
```

Comment:
Since we are on the topic, it is worth explaining these special characters. We can also just do this in the URL section

Reply:
The special character is in the web page content, it is not part of the URL

Reply:
I meant in URL search parameters you will see whitespace encoded as %20 and similar